queries/perl/highlights.scm

```diff
@@ -1,198 +1,89 @@
-  (scientific_notation)
-] @float
+(string_literal) @string
+(interpolated_string_literal) @string
+(quoted_word_list) @string
+(command_string) @string
+[(heredoc_token) (command_heredoc_token)] @string.special
+(heredoc_content) @string
+(heredoc_end) @string.special
+[(escape_sequence) (escaped_delimiter)] @string.special
 
-; version sortof counts as a kind of multipart integer
-(version) @constant
+[(quoted_regexp) (match_regexp)] @string.regexp
 
-; Package names are types
-(package_name) @type
+(autoquoted_bareword _?) @string.special
 
-; The special SUPER:: could be called a namespace. It isn't really but it
-;   should highlight differently and we might as well do it this way
-(super) @namespace
+[(scalar) (arraylen)] @variable.scalar
+(scalar_deref_expression ["->" "$" "*"] @variable.scalar)
+(array) @variable.array
+(array_deref_expression ["->" "@" "*"] @variable.array)
+(hash) @variable.hash
+(hash_deref_expression ["->" "%" "*"] @variable.hash)
 
-; Comments are comments
-(comments) @comment @spell
+(array_element_expression [array:(_) "->" "[" "]"] @variable.array)
+(slice_expression [array:(_) "->" "[" "]"] @variable.array)
+(keyval_expression [array:(_) "->" "[" "]"] @variable.array)
 
-((source_file . (comments) @preproc)
-  (#lua-match? @preproc "^#!/"))
+(hash_element_expression [hash:(_) "->" "{" "}"] @variable.hash)
+(slice_expression [hash:(_) "->" "[" "]"] @variable.hash)
+(keyval_expression [hash:(_) "->" "[" "]"] @variable.hash)
 
-; POD should be handled specially with its own embedded subtype but for now
-;   we'll just have to do this.
-(pod_statement) @text
+(hash_element_expression key: (bareword) @string.special)
 
-(method_invocation
-  function_name: (identifier) @method.call)
-(call_expression
-  function_name: (identifier) @function.call)
+(use_statement (package) @type)
+(package_statement (package) @type)
+(require_expression (bareword) @type)
 
-; Built-in functions
-((call_expression
-  function_name: (identifier) @function.builtin)
- (#any-of? @function.builtin
-   "print" "printf" "sprintf" "say"
-   "push" "pop" "shift" "unshift" "splice"
-   "exists" "delete" "keys" "values"
-   "each"))
+(subroutine_declaration_statement name: (_) @function)
+(attribute_name) @decorator
+(attribute_value) @string
 
-[
-  (map)
-  (grep)
-  (bless)
-] @function.builtin
+(goto_expression (label) @label)
+(loopex_expression (label) @label)
 
-;; ----------
+(statement_label label: _ @label)
 
-(use_constant_statement
-  constant: (identifier) @constant)
+(relational_expression operator: "isa" right: (bareword) @type)
 
-(named_block_statement
-  function_name: (identifier) @function)
+(function_call_expression (function) @function)
+(method_call_expression (method) @function.method)
+(method_call_expression invocant: (bareword) @type)
 
-(function_definition
-  name: (identifier) @function)
+(func0op_call_expression function: _ @function.builtin)
+(func1op_call_expression function: _ @function.builtin)
```

Comment:
Was there any specific reason for this weird syntax node? Or is it having separated nodes for these leads to more efficient parsing?

Reply:
I assume that's related to [this](https://www.perlmonks.org/?node_id=663393) quirk of Perl.

> ```perl
> whatever  / 25 ; # / ; die "this dies!";
> ```
>
> Schwartz's Snippet can parse two different ways: if `whatever` is nullary (that is, takes no arguments), the first statement is a division in void context, and the rest of the line is a comment. If `whatever` takes an argument, Schwartz's Snippet parses as a call to the `whatever` function with the result of a match operator, then a call to the `die()` function.

Reply:
> Was there any specific reason for this weird syntax node? Or is it having separated nodes for these leads to more efficient parsing?

Short answer: Because perl grammar is crazy.

Longer answer: Perl's parser (actually the lexer - yes really the lexer) categorizes the names of known functions into three separate sets - nullary functions, unary functions, and anything else. The parsing rules for these different categories take account of zero, at-most one or any number of arguments. Because of this, parsing of functions known to be in various categories has different precedence even without parens:

```perl
join $x, $y, $z   # is a regular list function, parses like  join($x, $y, $x)
int $x, $y, $z   # int is known to be a unary function, so this parses like  int($x), $y, $z
```

These rules don't matter for purely syntax-highlighting reasons, so previous technologies like `syntax/perl.vim` didn't have to care. But tooling that attempts to use the tree-sitter node tree result as some sort of AST and try to perform real code analysis will need to know about it, so it makes sense to attmept it here.

Reply:
TIL more about perl